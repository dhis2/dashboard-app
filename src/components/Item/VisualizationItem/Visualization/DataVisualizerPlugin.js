import PropTypes from 'prop-types'
import React, { Suspense, useState, useEffect, useCallback } from 'react'
import { useUserSettings } from '../../../UserSettingsProvider'
import LoadingMask from './LoadingMask'
import classes from './styles/DataVisualizerPlugin.module.css'
import VisualizationErrorMessage from './VisualizationErrorMessage'

const VisualizationPlugin = React.lazy(() =>
    import(
        /* webpackChunkName: "data-visualizer-plugin" */ /* webpackPrefetch: true */ '@dhis2/data-visualizer-plugin'
    )
)

const DataVisualizerPlugin = ({
    filterVersion,
    item,
    style,
    visualization,
    dashboardMode,
}) => {
    const { userSettings } = useUserSettings()
    const [visualizationLoaded, setVisualizationLoaded] = useState(false)
    const [error, setError] = useState(false)

    const onLoadingComplete = useCallback(
        () => setVisualizationLoaded(true),
        []
    )

    const onError = () => setError(true)

    useEffect(() => {
        setError(false)
    }, [filterVersion, visualization.type])

    if (error) {
        return (
            <div style={style}>
                <VisualizationErrorMessage
                    item={item}
                    dashboardMode={dashboardMode}
                />
            </div>
        )
    }

    return (
        <Suspense fallback={<div />}>
            {!visualizationLoaded && <LoadingMask style={style} />}
            <div className={classes.wrapper}>
                <VisualizationPlugin
                    visualization={visualization}
                    forDashboard={true}
                    userSettings={userSettings}
                    style={style}
                    onLoadingComplete={onLoadingComplete}
                    onError={onError}
                />
            </div>
        </Suspense>
    )
}

DataVisualizerPlugin.propTypes = {
    dashboardMode: PropTypes.string,
    filterVersion: PropTypes.string,
    item: PropTypes.object,
    style: PropTypes.object,
    visualization: PropTypes.object,
}

export default DataVisualizerPlugin
