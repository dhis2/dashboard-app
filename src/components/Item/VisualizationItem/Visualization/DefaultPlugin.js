import { useConfig } from '@dhis2/app-runtime'
import { useD2 } from '@dhis2/app-runtime-adapter-d2'
import PropTypes from 'prop-types'
import React, { useEffect, useRef } from 'react'
import getVisualizationContainerDomId from '../getVisualizationContainerDomId.js'
import { load, unmount } from './plugin.js'

const DefaultPlugin = ({
    item,
    activeType,
    filterVersion,
    mapViewCount,
    visualization,
    options,
    style,
}) => {
    const { d2 } = useD2()
    const { baseUrl } = useConfig()
    const credentials = {
        baseUrl,
        auth: d2.Api.getApi().defaultHeaders.Authorization,
    }

    const prevItem = useRef()
    const prevActiveType = useRef()
    const prevFilterVersion = useRef()
    const prevMapViewCount = useRef()
    const prevVisualization = useRef()

    useEffect(() => {
        load(item, visualization, {
            credentials,
            activeType,
            d2,
            options,
        })

        prevItem.current = item
        prevActiveType.current = activeType
        prevFilterVersion.current = filterVersion
        prevMapViewCount.current = mapViewCount
        prevVisualization.current = visualization

        return () => unmount(item, item.type || activeType)
    }, [])

    useEffect(() => {
        if (
            prevItem.current === item &&
            (prevActiveType.current !== activeType ||
                prevFilterVersion.current !== filterVersion ||
                prevMapViewCount.current < mapViewCount ||
                prevVisualization.current !== visualization)
        ) {
            /* Item is the same but type or filters has changed
             * or map was previously loaded with fewer mapViews
             * so necessary to reload
             */

            load(item, visualization, {
                credentials,
                activeType,
                d2,
                options,
            })
        }

        prevItem.current = item
        prevActiveType.current = activeType
        prevFilterVersion.current = filterVersion
        prevVisualization.current = visualization
    }, [item, visualization, activeType, filterVersion, mapViewCount])

    return <div id={getVisualizationContainerDomId(item.id)} style={style} />
}

DefaultPlugin.propTypes = {
    activeType: PropTypes.string,
    filterVersion: PropTypes.string,
    item: PropTypes.object,
    mapViewCount: PropTypes.number,
    options: PropTypes.object,
    style: PropTypes.object,
    visualization: PropTypes.object,
}

DefaultPlugin.defaultProps = {
    style: {},
    item: {},
    options: {},
    visualization: {},
}

export default DefaultPlugin
