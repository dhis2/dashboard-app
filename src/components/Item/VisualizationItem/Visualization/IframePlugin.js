import { useD2 } from '@dhis2/app-runtime-adapter-d2'
import postRobot from '@krakenjs/post-robot'
import PropTypes from 'prop-types'
import React, { useRef, useCallback, useState, useEffect } from 'react'
import { itemTypeMap } from '../../../../modules/itemTypes.js'
import { getPluginOverrides } from '../../../../modules/localStorage.js'
import { useCacheableSection } from '../../../../modules/useCacheableSection.js'
import { useUserSettings } from '../../../UserSettingsProvider.js'
import classes from './styles/DataVisualizerPlugin.module.css'
import VisualizationErrorMessage from './VisualizationErrorMessage.js'

const IframePlugin = ({
    filterVersion,
    item,
    style,
    visualization,
    dashboardMode,
    dashboardId,
}) => {
    const { d2 } = useD2()
    // TODO replace d2 with useCachedDataQuery
    const apps = d2.system.installedApps
    const { userSettings } = useUserSettings()
    const iframeRef = useRef()
    const [error, setError] = useState(false)

    // When this mounts, check if the dashboard is recording
    const { isCached, recordingState } = useCacheableSection(dashboardId)
    const [recordOnNextLoad, setRecordOnNextLoad] = useState(
        recordingState === 'recording'
    )

    const onError = () => setError(true)

    useEffect(() => {
        // Tell plugin to remove cached data if this dashboard has been removed
        // from offline storage
        if (iframeRef?.current && !isCached) {
            postRobot
                .send(iframeRef.current.contentWindow, 'removeCachedData')
                .catch((err) => {
                    // catch error if iframe hasn't loaded yet
                    const msg = 'No handler found for post message:'
                    if (err.message.startsWith(msg)) {
                        return
                    }
                    console.error(err)
                })
        }
    }, [isCached])

    useEffect(() => {
        if (iframeRef?.current) {
            const listener = postRobot.on(
                'getProps',
                // listen for messages coming only from the iframe rendered by this component
                { window: iframeRef.current.contentWindow },
                () => {
                    const pluginProps = {
                        isVisualizationLoaded: true,
                        forDashboard: true,
                        displayProperty: userSettings.displayProperty,
                        visualization,
                        onError,

                        // For caching: ---
                        // Add user & dashboard IDs to cache ID to avoid removing a cached
                        // plugin that might be used in another dashboard also
                        // TODO: May also want user ID too for multi-user situations
                        cacheId: `${dashboardId}-${item.id}`,
                        isParentCached: isCached,
                        recordOnNextLoad: recordOnNextLoad,
                    }

                    if (recordOnNextLoad) {
                        // Avoid recording unnecessarily,
                        // e.g. if plugin re-requests props for some reason
                        setRecordOnNextLoad(false)
                    }

                    return pluginProps
                }
            )

            return () => listener.cancel()
        }
    }, [
        visualization,
        userSettings,
        dashboardId,
        item.id,
        isCached,
        recordOnNextLoad,
    ])

    useEffect(() => {
        setError(false)
    }, [filterVersion, visualization.type])

    const getIframeSrc = useCallback(() => {
        // 1. check if there is an override for the plugin
        const pluginOverrides = getPluginOverrides()

        if (pluginOverrides && pluginOverrides[item.type]) {
            return pluginOverrides[item.type]
        }

        const appKey = itemTypeMap[item.type].appKey

        // 2. check if there is an installed app for the item type
        // and use its plugin launch URL
        if (appKey) {
            const appDetails = apps.find((app) => app.key === appKey)

            if (appDetails?.pluginLaunchUrl) {
                return appDetails.pluginLaunchUrl
            }
        }

        setError(true)

        return
    }, [item.type, apps])

    if (error) {
        return (
            <div style={style}>
                <VisualizationErrorMessage
                    item={item}
                    dashboardMode={dashboardMode}
                />
            </div>
        )
    }

    const iframeSrc = getIframeSrc()

    return (
        <div className={classes.wrapper}>
            {iframeSrc ? (
                <iframe
                    ref={iframeRef}
                    src={iframeSrc}
                    // preserve dimensions if provided
                    style={{
                        width: style.width || '100%',
                        height: style.height || '100%',
                        border: 'none',
                    }}
                ></iframe>
            ) : null}
        </div>
    )
}

IframePlugin.propTypes = {
    dashboardId: PropTypes.string,
    dashboardMode: PropTypes.string,
    filterVersion: PropTypes.string,
    item: PropTypes.object,
    style: PropTypes.object,
    visualization: PropTypes.object,
}

export default IframePlugin
