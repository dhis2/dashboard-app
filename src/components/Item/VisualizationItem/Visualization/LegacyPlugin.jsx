import { useConfig } from '@dhis2/app-runtime'
import PropTypes from 'prop-types'
import React, { useEffect, useRef } from 'react'
import getVisualizationContainerDomId from '../getVisualizationContainerDomId.js'
import { load, unmount } from './plugin.js'

const LegacyPlugin = ({
    item,
    activeType,
    visualization,
    style,
    gridWidth,
}) => {
    const { baseUrl } = useConfig()
    const prevItem = useRef()
    const prevActiveType = useRef()
    const prevVisualization = useRef()

    useEffect(() => {
        const el = document.querySelector(
            `#${getVisualizationContainerDomId(item.id)}`
        )
        if (typeof el?.setViewportSize === 'function') {
            setTimeout(
                () =>
                    el.setViewportSize(el.clientWidth - 5, el.clientHeight - 5),
                10
            )
        }
    }, [style.width, style.height, gridWidth, item.id])

    useEffect(() => {
        if (
            !prevItem.current ||
            (prevItem.current === item &&
                prevActiveType.current !== activeType) ||
            prevVisualization.current !== visualization
        ) {
            // Initial load, or active type or filter has changed
            load(item, visualization, {
                credentials: { baseUrl },
                activeType,
            })
        }

        prevItem.current = item
        prevActiveType.current = activeType
        prevVisualization.current = visualization

        return () => unmount(item, item.type || activeType)
    }, [item, visualization, activeType, baseUrl])

    return <div id={getVisualizationContainerDomId(item.id)} style={style} />
}

LegacyPlugin.propTypes = {
    item: PropTypes.object.isRequired,
    style: PropTypes.object.isRequired,
    visualization: PropTypes.object.isRequired,
    activeType: PropTypes.string,
    gridWidth: PropTypes.number,
}

export default LegacyPlugin
