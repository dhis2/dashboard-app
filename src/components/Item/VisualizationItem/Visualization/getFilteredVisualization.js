import { MAP } from '../../../../modules/itemTypes.js'

const mapViewIsThematicOrEvent = (mapView) =>
    mapView.layer.includes('thematic') || mapView.layer.includes('event')

const getFilteredMap = (visualization, filters) => {
    // apply filters only to thematic and event layers
    const mapViews = visualization.mapViews.map((mapView) => {
        if (mapViewIsThematicOrEvent(mapView)) {
            return getFilteredVisualization(mapView, filters)
        }

        return mapView
    })

    return {
        ...visualization,
        mapViews,
    }
}

const getFilteredVisualization = (visualization, filters, originalType) => {
    if (!Object.keys(filters).length) {
        return visualization
    }

    if (originalType === MAP) {
        return getFilteredMap(visualization, filters)
    }

    // deep clone objects in filters to avoid changing the visualization in the Redux store
    const visRows = visualization.rows.map((obj) => ({ ...obj }))
    const visColumns = visualization.columns.map((obj) => ({ ...obj }))
    const visFilters = visualization.filters.map((obj) => ({ ...obj }))

    Object.keys(filters).forEach((dimensionId) => {
        if (filters[dimensionId]) {
            let dimensionFound = false

            ;[visRows, visColumns, visFilters].forEach((dimensionObjects) =>
                dimensionObjects
                    .filter((obj) => obj.dimension === dimensionId)
                    .forEach((obj) => {
                        dimensionFound = true
                        obj.items = filters[dimensionId]
                    })
            )

            // add dimension to filters if not already present elsewhere
            if (!dimensionFound) {
                visFilters.push({
                    dimension: dimensionId,
                    items: filters[dimensionId],
                })
            }
        }
    })

    return {
        ...visualization,
        rows: visRows,
        columns: visColumns,
        filters: visFilters,
    }
}

export default getFilteredVisualization
