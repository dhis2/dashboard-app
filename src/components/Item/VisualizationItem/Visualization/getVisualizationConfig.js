import {
    VIS_TYPE_COLUMN,
    VIS_TYPE_PIVOT_TABLE,
    getAdaptedUiLayoutByType,
} from '@dhis2/analytics'
import { REPORT_TABLE, CHART, MAP } from '../../../../modules/itemTypes'

export const THEMATIC_LAYER = 'thematic'

const extractMapView = map =>
    map.mapViews && map.mapViews.find(mv => mv.layer.includes(THEMATIC_LAYER))

const getWithoutId = object => ({
    ...object,
    id: undefined,
})

const getVisualizationConfig = (visualization, originalType, activeType) => {
    if (originalType === MAP && originalType !== activeType) {
        const extractedMapView = extractMapView(visualization)

        if (extractedMapView === undefined) {
            return null
        }

        return getWithoutId({
            ...visualization,
            ...extractedMapView,
            mapViews: undefined,
            type: activeType === CHART ? VIS_TYPE_COLUMN : VIS_TYPE_PIVOT_TABLE,
        })
    } else if (originalType === REPORT_TABLE && activeType === CHART) {
        const layout = getAdaptedUiLayoutByType(visualization, VIS_TYPE_COLUMN)
        return getWithoutId({
            ...visualization,
            ...layout,
            type: VIS_TYPE_COLUMN,
        })
    } else if (originalType === CHART && activeType === REPORT_TABLE) {
        const layout = getAdaptedUiLayoutByType(
            visualization,
            VIS_TYPE_PIVOT_TABLE
        )
        return getWithoutId({
            ...visualization,
            ...layout,
            type: VIS_TYPE_PIVOT_TABLE,
        })
    }

    return getWithoutId(visualization)
}

export default getVisualizationConfig
