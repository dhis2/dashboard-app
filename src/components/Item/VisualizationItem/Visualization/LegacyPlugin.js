import PropTypes from 'prop-types'
import React, { useEffect } from 'react'
import getVisualizationContainerDomId from '../getVisualizationContainerDomId.js'
import DefaultPlugin from './DefaultPlugin.js'

const LegacyPlugin = ({
    availableHeight,
    availableWidth,
    gridWidth,
    ...props
}) => {
    useEffect(() => {
        const el = document.querySelector(
            `#${getVisualizationContainerDomId(props.item.id)}`
        )
        if (typeof el?.setViewportSize === 'function') {
            setTimeout(
                () =>
                    el.setViewportSize(el.clientWidth - 5, el.clientHeight - 5),
                10
            )
        }
    }, [availableHeight, availableWidth, gridWidth])

    return <DefaultPlugin {...props} />
}

LegacyPlugin.propTypes = {
    activeType: PropTypes.string,
    availableHeight: PropTypes.number,
    availableWidth: PropTypes.number,
    gridWidth: PropTypes.number,
    isFullscreen: PropTypes.bool,
    item: PropTypes.object,
}

export default LegacyPlugin
