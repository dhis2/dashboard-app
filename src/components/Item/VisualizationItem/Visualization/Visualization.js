import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import i18n from '@dhis2/d2-i18n'
import uniqueId from 'lodash/uniqueId'

import LegacyPlugin from './LegacyPlugin'
import MapPlugin from './MapPlugin'
import DataVisualizerPlugin from './DataVisualizerPlugin'
import NoVisualizationMessage from './NoVisualizationMessage'

import getFilteredVisualization from '../../../../modules/getFilteredVisualization'
import getVisualizationConfig from '../../../../modules/getVisualizationConfig'
import {
    VISUALIZATION,
    MAP,
    CHART,
    REPORT_TABLE,
} from '../../../../modules/itemTypes'
import { getVisualizationId } from '../../../../modules/item'
import memoizeOne from '../../../../modules/memoizeOne'
import { sGetVisualization } from '../../../../reducers/visualizations'
import {
    sGetItemFiltersRoot,
    DEFAULT_STATE_ITEM_FILTERS,
} from '../../../../reducers/itemFilters'
import { isEditMode } from '../../../Dashboard/dashboardModes'
import { pluginIsAvailable } from './plugin'

class Visualization extends React.Component {
    constructor(props) {
        super(props)

        this.memoizedGetFilteredVisualization = memoizeOne(
            getFilteredVisualization
        )
        this.memoizedGetVisualizationConfig = memoizeOne(getVisualizationConfig)

        this.getFilterVersion = memoizeOne(() => uniqueId())
    }

    render() {
        const {
            visualization,
            activeType,
            item,
            itemFilters,
            ...rest
        } = this.props

        if (!visualization) {
            return (
                <NoVisualizationMessage
                    message={i18n.t('No data to display')}
                />
            )
        }

        const style = { height: this.props.availableHeight }
        if (this.props.availableWidth) {
            style.width = this.props.availableWidth
        }

        const visualizationConfig = this.memoizedGetVisualizationConfig(
            visualization,
            item.type,
            activeType
        )

        const filterVersion = this.getFilterVersion(itemFilters)

        switch (activeType) {
            case VISUALIZATION:
            case CHART:
            case REPORT_TABLE: {
                return (
                    <DataVisualizerPlugin
                        visualization={this.memoizedGetFilteredVisualization(
                            visualizationConfig,
                            itemFilters
                        )}
                        style={style}
                        filterVersion={filterVersion}
                        item={item}
                        dashboardMode={this.props.dashboardMode}
                    />
                )
            }
            case MAP: {
                return (
                    <MapPlugin
                        item={item}
                        activeType={activeType}
                        visualization={visualizationConfig}
                        itemFilters={itemFilters}
                        applyFilters={this.memoizedGetFilteredVisualization}
                        filterVersion={filterVersion}
                        style={style}
                        {...rest}
                    />
                )
            }
            default: {
                return pluginIsAvailable(activeType || item.type) ? (
                    <LegacyPlugin
                        item={item}
                        activeType={activeType}
                        visualization={this.memoizedGetFilteredVisualization(
                            visualizationConfig,
                            itemFilters
                        )}
                        filterVersion={filterVersion}
                        style={style}
                        {...rest}
                    />
                ) : (
                    <NoVisualizationMessage
                        message={i18n.t(
                            'Unable to load the plugin for this item'
                        )}
                    />
                )
            }
        }
    }
}

Visualization.propTypes = {
    activeType: PropTypes.string,
    availableHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    availableWidth: PropTypes.number,
    dashboardMode: PropTypes.string,
    item: PropTypes.object,
    itemFilters: PropTypes.object,
    visualization: PropTypes.object,
}

const mapStateToProps = (state, ownProps) => {
    const itemFilters = !isEditMode(ownProps.dashboardMode)
        ? sGetItemFiltersRoot(state)
        : DEFAULT_STATE_ITEM_FILTERS

    return {
        itemFilters,
        visualization: sGetVisualization(
            state,
            getVisualizationId(ownProps.item)
        ),
    }
}

export default connect(mapStateToProps)(Visualization)
