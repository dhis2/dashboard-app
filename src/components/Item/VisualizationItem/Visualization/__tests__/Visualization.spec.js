import { render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import Visualization from '../Visualization.js'

jest.mock('@dhis2/analytics', () => ({
    useCachedDataQuery: () => ({
        currentUser: {
            username: 'rainbowDash',
            id: 'r3nb0d5h',
        },
    }),
}))

jest.mock(
    '../LegacyPlugin',
    () =>
        function MockLegacyPlugin() {
            return <div className="legacy-plugin" />
        }
)

jest.mock(
    '../IframePlugin',
    () =>
        function MockIframePlugin() {
            return <div className="iframe-plugin" />
        }
)

const mockStore = configureMockStore()
const DEFAULT_STORE_WITH_ONE_ITEM = {
    visualizations: { rainbowVis: { rows: [], columns: [], filters: [] } },
    itemFilters: {},
    selected: {
        id: 'test-dashboard',
    },
}

test('renders a VisualizationPlugin when activeType is MAP', () => {
    const { container } = render(
        <Provider store={mockStore(DEFAULT_STORE_WITH_ONE_ITEM)}>
            <Visualization
                item={{
                    id: 'rainbow',
                    type: 'MAP',
                    map: { id: 'rainbowVis' },
                }}
                activeType="MAP"
                itemFilters={{}}
                availableHeight={500}
            />
        </Provider>
    )
    expect(container).toMatchSnapshot()
})

test('renders a VisualizationPlugin for CHART', () => {
    const { container } = render(
        <Provider store={mockStore(DEFAULT_STORE_WITH_ONE_ITEM)}>
            <Visualization
                item={{
                    id: 'rainbow',
                    type: 'VISUALIZATION',
                    visualization: { id: 'rainbowVis', type: 'BAR' },
                }}
                activeType="CHART"
                itemFilters={{}}
                availableHeight={500}
            />
        </Provider>
    )
    expect(container).toMatchSnapshot()
})

test('renders a VisualizationPlugin for REPORT_TABLE', () => {
    const { container } = render(
        <Provider store={mockStore(DEFAULT_STORE_WITH_ONE_ITEM)}>
            <Visualization
                item={{
                    id: 'rainbow',
                    type: 'VISUALIZATION',
                    visualization: { id: 'rainbowVis', type: 'PIVOT_TABLE' },
                }}
                activeType="REPORT_TABLE"
                itemFilters={{}}
                availableHeight={500}
            />
        </Provider>
    )
    expect(container).toMatchSnapshot()
})

test('renders a DefaultPlugin when activeType is EVENT_CHART', () => {
    const { container } = render(
        <Provider store={mockStore(DEFAULT_STORE_WITH_ONE_ITEM)}>
            <Visualization
                item={{
                    id: 'rainbow',
                    type: 'EVENT_CHART',
                    eventChart: { id: 'rainbowVis' },
                }}
                activeType="EVENT_CHART"
                itemFilters={{}}
                availableHeight={500}
            />
        </Provider>
    )
    expect(container).toMatchSnapshot()
})

test('renders a DefaultPlugin when activeType is EVENT_REPORT', () => {
    const { container } = render(
        <Provider store={mockStore(DEFAULT_STORE_WITH_ONE_ITEM)}>
            <Visualization
                item={{
                    id: 'rainbow',
                    type: 'EVENT_REPORT',
                    eventReport: { id: 'rainbowVis' },
                }}
                activeType="EVENT_REPORT"
                itemFilters={{}}
                availableHeight={500}
            />
        </Provider>
    )
    expect(container).toMatchSnapshot()
})

test('renders NoVisMessage when no visualization', () => {
    const store = {
        visualizations: {},
        itemFilters: {},
        selected: {
            id: 'test-dashboard',
        },
    }
    const { container } = render(
        <Provider store={mockStore(store)}>
            <Visualization
                item={{
                    id: 'rainbow',
                    type: 'VISUALIZATION',
                    visualization: { type: 'BAR' },
                }}
                activeType="CHART"
                itemFilters={{}}
                availableHeight={500}
            />
        </Provider>
    )
    expect(container).toMatchSnapshot()
})
