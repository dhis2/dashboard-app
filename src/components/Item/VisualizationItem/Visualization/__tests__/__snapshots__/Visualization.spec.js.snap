// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders NoVisMessage when no visualization 1`] = `
<div>
  <div
    class="message"
  >
    No data to display
  </div>
</div>
`;

exports[`renders a DefaultPlugin when activeType is EVENT_CHART 1`] = `
<div>
  <defaultplugin
    activetype="EVENT_CHART"
    item="[object Object]"
    itemfilters="[object Object]"
    style="height: 500px;"
    visualization="[object Object]"
  />
</div>
`;

exports[`renders a DefaultPlugin when activeType is EVENT_REPORT 1`] = `
<div>
  <defaultplugin
    activetype="EVENT_REPORT"
    item="[object Object]"
    itemfilters="[object Object]"
    style="height: 500px;"
    visualization="[object Object]"
  />
</div>
`;

exports[`renders a MapPlugin when activeType is MAP 1`] = `
<div>
  <mapplugin
    activetype="MAP"
    item="[object Object]"
    itemfilters="[object Object]"
    style="height: 500px;"
    visualization="[object Object]"
  />
</div>
`;

exports[`renders a VisualizationPlugin for CHART 1`] = `
<div>
  <div
    style="height: 500px;"
  >
    <div
      class="center"
    >
      <div
        class="jsx-19995197 "
        data-test="dhis2-uicore-circularloader"
        role="progressbar"
      >
        <svg
          class="jsx-19995197"
          viewBox="22 22 44 44"
        >
          <circle
            class="jsx-19995197 circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
    </div>
  </div>
  <visualizationplugin
    style="height: 500px;"
    visualization="[object Object]"
  />
</div>
`;

exports[`renders a VisualizationPlugin for REPORT_TABLE 1`] = `
<div>
  <div
    style="height: 500px;"
  >
    <div
      class="center"
    >
      <div
        class="jsx-19995197 "
        data-test="dhis2-uicore-circularloader"
        role="progressbar"
      >
        <svg
          class="jsx-19995197"
          viewBox="22 22 44 44"
        >
          <circle
            class="jsx-19995197 circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
    </div>
  </div>
  <visualizationplugin
    style="height: 500px;"
    visualization="[object Object]"
  />
</div>
`;

exports[`renders active type MAP rather than original type REPORT_TABLE 1`] = `
<div>
  <mapplugin
    activetype="MAP"
    item="[object Object]"
    itemfilters="[object Object]"
    style="height: 500px;"
    visualization="[object Object]"
  />
</div>
`;

exports[`renders active type REPORT_TABLE rather than original type MAP 1`] = `
<div>
  <div
    style="height: 500px;"
  >
    <div
      class="center"
    >
      <div
        class="jsx-19995197 "
        data-test="dhis2-uicore-circularloader"
        role="progressbar"
      >
        <svg
          class="jsx-19995197"
          viewBox="22 22 44 44"
        >
          <circle
            class="jsx-19995197 circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
    </div>
  </div>
  <visualizationplugin
    style="height: 500px;"
  />
</div>
`;
